utxo/src/utxo_impl/simulation.rs

```diff
@@ -15,86 +15,69 @@
+};
+use crypto::random::Rng;
+use rstest::rstest;
+use test_utils::random::{make_seedable_rng, Seed};
 
 fn populate_cache<'a>(
+    rng: &mut impl Rng,
     parent: &'a UtxosCache,
-    size: u64,
+    new_utxo_count: u64,
     existing_outpoints: &[OutPoint],
 ) -> (UtxosCache<'a>, Vec<OutPoint>) {
     let mut cache = UtxosCache::new(parent);
-
     // tracker
-    let mut outps: Vec<OutPoint> = vec![];
+    let mut outpoints: Vec<OutPoint> = vec![];
 
     // let's add utxos based on `size`.
-    for i in 0..size {
-        let block_height = if random_bool() {
-            i
-        } else {
-            // setting a random height based on the `size`.
-            make_pseudo_rng().gen_range(0..size)
-        };
-
-        let (utxo, outpoint) = create_utxo(block_height);
+    for _ in 0..new_utxo_count {
+        let block_height = rng.gen_range(0..new_utxo_count);
+        let (utxo, outpoint) = create_utxo(rng, block_height);
 
-        let outpoint = if random_bool() && existing_outpoints.len() > 1 {
+        let outpoint = if rng.gen::<bool>() && existing_outpoints.len() > 1 {
             // setting a random existing 'spent' outpoint
-            let rng = make_pseudo_rng().gen_range(0..existing_outpoints.len());
-            let outpoint = &existing_outpoints[rng];
-
-            outpoint.clone()
+            let outpoint_idx = rng.gen_range(0..existing_outpoints.len());
+            existing_outpoints[outpoint_idx].clone()
         } else {
             // tracking the outpoints
-            outps.push(outpoint.clone());
+            outpoints.push(outpoint.clone());
             outpoint
         };
 
         // randomly set the `possible_overwrite`
-        let possible_overwrite = random_bool();
+        let possible_overwrite = rng.gen::<bool>();
         let _ = cache.add_utxo(&outpoint, utxo, possible_overwrite);
-
-        // println!("child, insert: {:?}, overwrite: {}", outpoint,possible_overwrite );
     }
 
     // let's create half of the outpoints provided, to be marked as spent.
     // there's a possibility when randomly the same outpoint is used, so half seems okay.
-    let spent_size = outps.len() / 2;
+    let spent_size = outpoints.len() / 2;
 
     for _ in 0..spent_size {
         // randomly select which outpoint should be marked as "spent"
-        if random_bool() && existing_outpoints.len() > 1 {
+        if rng.gen::<bool>() && existing_outpoints.len() > 1 {
             // just call the `spend_utxo`. Does not matter if it removes the outpoint entirely,
             // or just mark it as `spent`,
-            let outp_idx = make_pseudo_rng().gen_range(0..existing_outpoints.len());
+            let outp_idx = rng.gen_range(0..existing_outpoints.len());
             let to_spend = &existing_outpoints[outp_idx];
-            assert!(cache.spend_utxo(to_spend).is_ok());
-
-            //println!("child, spend: {:?}, removed", to_spend);
+            let _ = cache.spend_utxo(to_spend);
```

Comment:
The fix is basically the removal of this assertion. I see no value in panicing if we try to spend a utxo but it was already spent. This happens pretty frequently because utxo is chosen randomly from existing ones that could already be spent. Moreover, the purpose of this function is to "populate cache" - to make some random chaotic actions.